### PR TITLE
Replace util/webhook code with CustomDefaulter/CustomValidator

### LINKED ISSUE #3103: Replace util/webhook code with CustomDefaulter/CustomValidator
/kind feature

**Describe the solution you'd like**

@sbueringer [pointed out](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3080#discussion_r1087679502) that much of the CAPZ webhook utility code in `util/webhook` could possibly be replaced by using CustomDefaulter and CustomValidator from controller-runtime. Currently only `azuremachinetemplate_webhook.go` sets up its webhook this way. (But not fully--see the comment below.)

**Anything else you would like to add:**

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by willie-yao**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR refactors existing webhooks that require the use of a `client`. Instead of registering the webhooks manually and deviating from the general webhook pattern, we use a CustomDefaulter/CustomValidator to provide the webhook with the `client`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3103

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by nojnhuh — COMMENTED** [review]
Left a few comments inline that apply to all of the webhooks, but nothing I'd block this PR on.

File: api/v1beta1/azuremachine_webhook.go
Comment (willie-yao) [reply api/v1beta1/azuremachine_webhook.go:51]: Ah that makes sense. Do you mean that it's not necessary to include these lines because the calls to `WithDefaulter` and `WithValidator` accomplish the same thing?

**Review by nojnhuh — APPROVED** [review]
/lgtm

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Comment by CecileRobertMichon** [comment]
/hold

let's wait until we cut the v1.8.0 tag before merging

**Comment by CecileRobertMichon** [comment]
/retest

e2e should be fixed now

**Pull request merged**
